internal/server/auditsink/auditsink.go

```diff
@@ -0,0 +1,204 @@
+			Key:   "flipt.version",
+			Value: attribute.StringValue(ae.Version),
+		})
+	}
+
+	if ae.Metadata.Action != "" {
+		akv = append(akv, attribute.KeyValue{
+			Key:   "flipt.metadata.action",
+			Value: attribute.StringValue(string(ae.Metadata.Action)),
+		})
+	}
+
+	if ae.Metadata.Type != "" {
+		akv = append(akv, attribute.KeyValue{
+			Key:   "flipt.metadata.type",
+			Value: attribute.StringValue(string(ae.Metadata.Type)),
+		})
+	}
+
+	if ae.Payload != nil {
+		b, _ := json.Marshal(ae.Payload)
+
+		akv = append(akv, attribute.KeyValue{
+			Key:   "flipt.payload",
+			Value: attribute.StringValue(string(b)),
+		})
+	}
+
+	return akv
+}
+
+func deserializeHelper(s string) string {
+	ss := strings.Split(s, ".")
+	return ss[len(ss)-1]
+}
+
+func (ae *AuditEvent) Valid() bool {
+	return ae.Version != "" && ae.Metadata.Action != "" && ae.Metadata.Type != "" && ae.Payload != nil
+}
+
+// decodeToAuditEvent provides helper logic for turning to value of SpanEvents to
+// an AuditEvent.
+func decodeToAuditEvent(kvs []attribute.KeyValue) (*AuditEvent, error) {
+	ae := new(AuditEvent)
+	for _, kv := range kvs {
+		terminalKey := deserializeHelper(string(kv.Key))
+
+		if terminalKey == "version" {
+			ae.Version = kv.Value.AsString()
+		}
+
+		if terminalKey == "action" {
+			ae.Metadata.Action = AuditAction(kv.Value.AsString())
+		}
+
+		if terminalKey == "type" {
+			ae.Metadata.Type = AuditType(kv.Value.AsString())
+		}
+
+		if terminalKey == "payload" {
+			var payload interface{}
+			if err := json.Unmarshal([]byte(kv.Value.AsString()), &payload); err != nil {
+				return nil, err
+			}
+			ae.Payload = payload
+		}
+	}
+
+	if !ae.Valid() {
+		return nil, fmt.Errorf("audit event not valid")
+	}
+
+	return ae, nil
+}
+
+// MetadataConfig holds information of what metadata an event will contain.
+type MetadataConfig struct {
+	Type   AuditType   `json:"type"`
+	Action AuditAction `json:"action"`
+}
+
+// AuditSink is the abstraction for various audit sink configurations
+// that Flipt will support.
+type AuditSink interface {
+	SendAudits([]AuditEvent) error
+	fmt.Stringer
+}
+
+// SinkSpanExporter sends audit logs to configured sinks through intercepting span events.
+type SinkSpanExporter struct {
+	sinks  []AuditSink
+	logger *zap.Logger
+}
+
+// AuditEventExporter provides an API for exporting spans as AuditEvent(s).
+type AuditEventExporter interface {
+	ExportSpans(ctx context.Context, spans []trace.ReadOnlySpan) error
+	Shutdown(ctx context.Context) error
+	SendAudits(aes []AuditEvent) error
+}
+
+// NewSinkSpanExporter is the constructor for a SinkSpanExporter.
+func NewSinkSpanExporter(logger *zap.Logger, sinks []AuditSink) AuditEventExporter {
+	return &SinkSpanExporter{
+		sinks:  sinks,
+		logger: logger,
+	}
+}
+
+// ExportSpans completes one part of the implementation of a SpanExporter. Decodes span events to audit events.
+func (s *SinkSpanExporter) ExportSpans(ctx context.Context, spans []trace.ReadOnlySpan) error {
```

Comment:
Right now this is synchronous.. We can look at making the content of this asynchronous. Pending benchmarks/load tests

Reply:
I could be wrong, but it might not be a bad idea to keep this synchronous as you have it here (indefinitely).

Given the trace provider handles batching of events, we might get a more predictable experience if this type produces back-pressure to the otel SDK.
Need to dig into it more though to understand that better, but a cursory glance shows that the batch processor is handling queueing and flushing (unsurprisingly) asynchronously: https://github.com/open-telemetry/opentelemetry-go/blob/main/sdk/trace/batch_span_processor.go#L125-L130